{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23105735806966033\n"
     ]
    }
   ],
   "source": [
    "# Part 1 : Bioinformatics Stronghold: \n",
    "## Assignment: Nov 17th\n",
    "# FIBD (Fibonacci Rabbits dying)\n",
    "with open('rosalind_fibd.txt') as input_data:\n",
    "    n,m = map(int, input_data.read().split())\n",
    "#initialization ( first rabbits)\n",
    "Rabbits = [1]+[0]*(m-1)\n",
    "# initial population in year 0\n",
    "# baby rabbits are bunnies (new rabbits)\n",
    "for year in range(1, n):\n",
    "    Bunnies = 0\n",
    "    # Get the number of Rabbits able to old enough to give birth.\n",
    "    for j in range(1,m):\n",
    "        Bunnies += Rabbits[(year-j-1)%m]\n",
    "    # Bunnies replace the old rabbits who died.\n",
    "    Rabbits[(year)%m] = Bunnies\n",
    "Total_Rabbits = sum(Rabbits)\n",
    "print(Total_Rabbits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "193664\n"
     ]
    }
   ],
   "source": [
    "#mRNA (Inferring mRNA from Protein)\n",
    "\n",
    "from Bio.Data.CodonTable import standard_rna_table\n",
    "\n",
    "# Load the standard codon table\n",
    "codon_table = standard_rna_table.forward_table\n",
    "stop_codons = standard_rna_table.stop_codons\n",
    "\n",
    "# Create a dictionary to count codons per amino acid\n",
    "rna_dict = {}\n",
    "\n",
    "# Populate rna_dict with counts of codons for each amino acid\n",
    "for codon, amino_acid in codon_table.items():\n",
    "    if amino_acid not in rna_dict:\n",
    "        rna_dict[amino_acid] = 0\n",
    "    rna_dict[amino_acid] += 1\n",
    "\n",
    "# Add stop codons to the dictionary\n",
    "rna_dict['Stop'] = len(stop_codons)\n",
    "\n",
    "with open('rosalind_mrna.txt') as input_data:\n",
    "    protein = input_data.read().strip()\n",
    "\n",
    "# Initialize with the number of stop codons\n",
    "rna_num = rna_dict['Stop']\n",
    "\n",
    "# Multiply by the number of codons for each amino acid in the protein\n",
    "for p in protein:\n",
    "    rna_num = (rna_num * rna_dict[p]) % 1000000\n",
    "\n",
    "print(rna_num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.4333521110224611\n"
     ]
    }
   ],
   "source": [
    "# LIA (Independent Alleles/Mendell's Second Law)\n",
    "from math import comb\n",
    "#Computes the number of ways to choose i successful Aa Bb individuals from total_offspring\n",
    "with open('rosalind_lia.txt') as input_data:\n",
    "    k, N = map(int, input_data.read().split())\n",
    "\n",
    "prob = 0\n",
    "total_offspring = 2**k\n",
    "\n",
    "for i in range(N, total_offspring + 1):\n",
    "    prob += comb(total_offspring, i) * ((1/4.0)**i) * ((3/4.0)**(total_offspring - i))\n",
    "\n",
    "print(prob)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "111075.94424\n"
     ]
    }
   ],
   "source": [
    "with open('rosalind_prtm.txt') as file:\n",
    "    protein_str = file.read().strip()\n",
    "# Monoisotopic mass table from wiki\n",
    "monoisotopic_mass = {\n",
    "    'A': 71.03711, 'C': 103.00919, 'D': 115.02694, 'E': 129.04259,\n",
    "    'F': 147.06841, 'G': 57.02146, 'H': 137.05891, 'I': 113.08406,\n",
    "    'K': 128.09496, 'L': 113.08406, 'M': 131.04049, 'N': 114.04293,\n",
    "    'P': 97.05276, 'Q': 128.05858, 'R': 156.10111, 'S': 87.03203,\n",
    "    'T': 101.04768, 'V': 99.06841, 'W': 186.07931, 'Y': 163.06333,\n",
    "}\n",
    "total_weight = sum(monoisotopic_mass[aa] for aa in protein_str)\n",
    "print(total_weight)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATCTAGTACGGACCCAGATTCGGATTTCTAAATTCCTCGGTATCACTGCCGGGGCAACCGGCCTCTTTGGAGGTACGCGGCTCAGTCCTAATTGGCGTTTATAGAGGATCGCTTCTACTAGAACGCGCTAACTGCGTATTACTCGGCACGGAGGGCTGTGTCACTGGGGGTGC\n"
     ]
    }
   ],
   "source": [
    "#Lcsm (Finding a Shared Motif)\n",
    "# Read a FASTA file and return a list of sequences\n",
    "\n",
    "def ReadFASTA(file_path):\n",
    "    with open(file_path, 'r') as file:\n",
    "        sequences = []\n",
    "        current_seq = []\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if line.startswith('>'):\n",
    "                if current_seq:\n",
    "                    sequences.append(''.join(current_seq))\n",
    "                current_seq = []\n",
    "            else:\n",
    "                current_seq.append(line)\n",
    "        if current_seq:\n",
    "            sequences.append(''.join(current_seq))\n",
    "    return sequences\n",
    "\n",
    "#Finds the longest common substring in a list of DNA sequences.\n",
    "def LongestCommonSubstring(sequences):\n",
    "\n",
    "    reference = sequences[0]  # Use the first sequence as the reference\n",
    "    longest_substring = ''\n",
    "    \n",
    "    for i in range(len(reference)):\n",
    "        for j in range(i + 1, len(reference) + 1):\n",
    "            candidate = reference[i:j]\n",
    "            if all(candidate in seq for seq in sequences):\n",
    "                if len(candidate) > len(longest_substring):\n",
    "                    longest_substring = candidate\n",
    "    return longest_substring\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    sequences = ReadFASTA('rosalind_lcsm.txt')\n",
    "    result = LongestCommonSubstring(sequences)\n",
    "    print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "120\n",
      "1 2 3 4 5\n",
      "1 2 3 5 4\n",
      "1 2 4 3 5\n",
      "1 2 4 5 3\n",
      "1 2 5 3 4\n",
      "1 2 5 4 3\n",
      "1 3 2 4 5\n",
      "1 3 2 5 4\n",
      "1 3 4 2 5\n",
      "1 3 4 5 2\n",
      "1 3 5 2 4\n",
      "1 3 5 4 2\n",
      "1 4 2 3 5\n",
      "1 4 2 5 3\n",
      "1 4 3 2 5\n",
      "1 4 3 5 2\n",
      "1 4 5 2 3\n",
      "1 4 5 3 2\n",
      "1 5 2 3 4\n",
      "1 5 2 4 3\n",
      "1 5 3 2 4\n",
      "1 5 3 4 2\n",
      "1 5 4 2 3\n",
      "1 5 4 3 2\n",
      "2 1 3 4 5\n",
      "2 1 3 5 4\n",
      "2 1 4 3 5\n",
      "2 1 4 5 3\n",
      "2 1 5 3 4\n",
      "2 1 5 4 3\n",
      "2 3 1 4 5\n",
      "2 3 1 5 4\n",
      "2 3 4 1 5\n",
      "2 3 4 5 1\n",
      "2 3 5 1 4\n",
      "2 3 5 4 1\n",
      "2 4 1 3 5\n",
      "2 4 1 5 3\n",
      "2 4 3 1 5\n",
      "2 4 3 5 1\n",
      "2 4 5 1 3\n",
      "2 4 5 3 1\n",
      "2 5 1 3 4\n",
      "2 5 1 4 3\n",
      "2 5 3 1 4\n",
      "2 5 3 4 1\n",
      "2 5 4 1 3\n",
      "2 5 4 3 1\n",
      "3 1 2 4 5\n",
      "3 1 2 5 4\n",
      "3 1 4 2 5\n",
      "3 1 4 5 2\n",
      "3 1 5 2 4\n",
      "3 1 5 4 2\n",
      "3 2 1 4 5\n",
      "3 2 1 5 4\n",
      "3 2 4 1 5\n",
      "3 2 4 5 1\n",
      "3 2 5 1 4\n",
      "3 2 5 4 1\n",
      "3 4 1 2 5\n",
      "3 4 1 5 2\n",
      "3 4 2 1 5\n",
      "3 4 2 5 1\n",
      "3 4 5 1 2\n",
      "3 4 5 2 1\n",
      "3 5 1 2 4\n",
      "3 5 1 4 2\n",
      "3 5 2 1 4\n",
      "3 5 2 4 1\n",
      "3 5 4 1 2\n",
      "3 5 4 2 1\n",
      "4 1 2 3 5\n",
      "4 1 2 5 3\n",
      "4 1 3 2 5\n",
      "4 1 3 5 2\n",
      "4 1 5 2 3\n",
      "4 1 5 3 2\n",
      "4 2 1 3 5\n",
      "4 2 1 5 3\n",
      "4 2 3 1 5\n",
      "4 2 3 5 1\n",
      "4 2 5 1 3\n",
      "4 2 5 3 1\n",
      "4 3 1 2 5\n",
      "4 3 1 5 2\n",
      "4 3 2 1 5\n",
      "4 3 2 5 1\n",
      "4 3 5 1 2\n",
      "4 3 5 2 1\n",
      "4 5 1 2 3\n",
      "4 5 1 3 2\n",
      "4 5 2 1 3\n",
      "4 5 2 3 1\n",
      "4 5 3 1 2\n",
      "4 5 3 2 1\n",
      "5 1 2 3 4\n",
      "5 1 2 4 3\n",
      "5 1 3 2 4\n",
      "5 1 3 4 2\n",
      "5 1 4 2 3\n",
      "5 1 4 3 2\n",
      "5 2 1 3 4\n",
      "5 2 1 4 3\n",
      "5 2 3 1 4\n",
      "5 2 3 4 1\n",
      "5 2 4 1 3\n",
      "5 2 4 3 1\n",
      "5 3 1 2 4\n",
      "5 3 1 4 2\n",
      "5 3 2 1 4\n",
      "5 3 2 4 1\n",
      "5 3 4 1 2\n",
      "5 3 4 2 1\n",
      "5 4 1 2 3\n",
      "5 4 1 3 2\n",
      "5 4 2 1 3\n",
      "5 4 2 3 1\n",
      "5 4 3 1 2\n",
      "5 4 3 2 1\n"
     ]
    }
   ],
   "source": [
    "#PERM (Enumerating Gene Orders)\n",
    "\n",
    "from itertools import permutations\n",
    "from math import factorial\n",
    "\n",
    "with open('rosalind_perm.txt') as input_file:\n",
    "    n = int(input_file.read().strip())\n",
    "\n",
    "# Generate all permutations of numbers 1 through n via permutation\n",
    "perm_list = permutations(range(1, n + 1))\n",
    "\n",
    "# Write to the output file each in in seperate line and also divided by space\n",
    "with open('output_perm.txt', 'w') as output_file:\n",
    "    # Write the total number of permutations (n!)\n",
    "    output_file.write(f\"{factorial(n)}\\n\")\n",
    "    \n",
    "print(factorial(n))\n",
    "for perm in permutations(range(1, n + 1)):\n",
    "    print(' '.join(map(str, perm)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGGTCAAAGTCACCATACATATAGTCGATAGCGGATTGACATAGACAGGCCAAGCAGCCAACCATTGGCATCTTATGCGCTAGATCCAGACTTCAACGAAACATTGTCTTGAAACCTCAAAGTATACCAACGAACCGAAACGCAAAGGCGTCCTATACAAAAACAAGAAACCTAAGGAGTTCCAAGTACTATTGTTGCTCAATGGGCCCACCTGAATCCGGTGGACCGGGATATAGTCGGATTAGCCCGAATAAATCGTATCCGCACACTGCATATATCCACCGACCGAGAAGGAGCCTGAGCCCGTAATGCAAAAAAAACAAAGTGAGTCGTTCGGATGAAAAAGGGATTAGATACGCCGTTAGTAAAACTCACACGTTTCTCCAAGTGGTACCATAGCAGGGAGTAACAGACTGAACAGAGGTCATACGTATATATGACAAACCAACACCTCTCAGTCACAGAACTATTGACGACGCGCACCCCAGTAAGAACGCTGGCCACGCGGACATCAGCCGAGAAGCCTCCGAACCTATGGGTCTTTAACCGTTGATTCATAAAATTAACGCGATGCGACATAATCGGACCCGGTACAACACAGCTTGAGCGGTAGATGATATCAGATAAAACCCTGGACTTAAAGCGGTACTACACCAATATACATGTCCACACAGCGGGCAAAATTAAGTACAATCATTCTTGGACACATTTACCCATCAATTAAGCATCGACCCCCAGGCCGGACGGAAGAGTCCCCAGAAAGCATTATAAATGAAACCAGAATAAAATAGACGGTCTGAAAGGCGATAATGAGCAAATGCAAATCGCGCATCGGAAACATTCCAAACCATATTACGCACTACAAGTGAGAACCAATTCAACATGAGGCCACTGGAAAAGATATCAACAGTCCTCCGGAATGCAGGGAATTCTCACATACGGAAAAAAGCCATG\n",
      "A: 3 2 0 2 2 4 5 4 3 0 1 4 1 2 3 2 4 2 3 0 4 1 3 2 3 1 2 4 2 4 4 2 1 1 4 0 1 2 5 1 4 3 7 2 6 0 4 2 3 2 1 6 4 0 2 3 1 2 0 4 3 2 2 4 3 2 2 4 2 5 3 3 3 0 5 3 2 2 1 2 2 4 3 3 2 3 3 4 2 4 0 1 2 0 4 4 2 2 4 5 5 2 6 2 0 3 2 3 3 0 0 4 3 4 2 0 2 0 6 5 4 3 2 4 2 4 1 3 3 4 1 3 3 4 1 1 1 3 5 4 1 2 1 3 5 3 2 3 2 2 3 0 3 3 7 1 4 0 4 4 5 3 5 3 3 5 1 4 4 3 0 1 2 3 4 1 2 3 2 2 2 2 3 6 4 2 3 3 2 2 4 2 2 2 2 0 3 1 2 2 5 5 2 1 2 1 2 0 1 5 2 0 1 0 4 3 1 0 1 2 2 2 0 2 4 1 1 1 2 2 4 2 3 3 4 2 2 2 2 3 4 2 3 4 2 3 2 2 3 3 6 1 4 4 4 2 3 2 2 3 2 2 2 1 4 4 3 6 1 3 2 3 4 2 4 0 3 3 3 1 5 2 1 1 5 2 2 1 4 0 7 3 2 2 3 3 2 2 3 3 6 2 2 2 2 2 3 5 5 1 1 2 3 4 5 3 5 4 3 3 2 4 4 4 2 3 2 3 2 0 1 1 3 3 1 1 1 4 2 2 3 5 4 3 3 1 1 2 5 2 1 4 3 3 2 3 1 3 2 3 2 1 2 3 3 3 3 4 3 5 2 0 1 4 2 5 2 1 3 2 1 2 2 2 3 4 3 2 0 3 1 2 5 1 2 4 1 3 3 1 4 1 2 1 4 1 2 5 5 1 3 3 3 2 2 1 4 3 2 3 1 4 4 2 1 1 8 2 4 3 0 2 3 0 3 2 4 1 0 3 2 4 7 6 1 3 4 3 2 4 4 1 0 2 2 3 4 2 3 3 4 0 6 2 5 3 2 0 5 3 2 2 3 2 2 5 1 3 2 2 2 4 1 1 2 1 3 1 1 4 4 3 3 4 2 1 2 2 2 1 3 2 5 1 2 1 3 3 5 2 4 2 1 5 3 3 0 1 4 2 4 4 0 2 0 3 2 2 2 3 4 3 4 1 3 2 2 0 2 1 0 2 1 1 5 3 2 2 2 3 2 3 4 3 1 1 6 1 6 6 3 6 3 1 3 3 1 3 3 2 4 2 2 2 2 5 2 5 2 4 3 0 2 1 0 5 1 0 3 0 0 3 3 1 3 4 1 4 2 3 2 1 2 3 0 3 3 0 1 3 1 5 2 4 3 2 4 1 6 1 1 5 1 5 2 5 4 3 3 1 0 1 1 1 1 4 2 2 1 5 4 3 3 1 2 2 2 5 0 2 4 1 5 3 2 4 3 1 4 2 4 2 4 2 3 1 1 2 4 3 4 2 4 3 1 2 3 1 3 4 3 5 5 2 0 6 3 2 3 4 1 4 3 3 3 4 2 2 2 1 1 2 2 4 3 4 2 4 2 3 1 5 2 2 1 4 0 2 3 3 2 3 3 4 3 1 5 1 1 2 4 3 1 3 2 3 4 2 3 1 2 2 2 5 2 2 2 3 6 2 4 2 2 1 2 2 1 4 2 3 3 5 2 2 5 2 1 5 3 4 3 5 2 1 4 5 5 3 1 3 2 3 4 0 5 4 6 4 3 4 2 5 1 2 1 1 1 0 3 4 4 6 1 1 3 0 4 2 3 4 2 1 3 2 0 3 3 3 1 0 2 4 4 3 2 4 0 2 2 3 3 3 2 2 2 4 6 4 2 5 1 2 2 2 3 3 3 1 3 4 0 5 2 1 3 3 3 2 4 1 2 4 2 4 4 3 2 3 3 1 5 4 0 3 5 5 2 1 2 3 5 2 5 3 1 5 3 1 2 1 4 2 1 1 1 3 3 3 3 0 3 2 5 3 1 4 4 1 4 0 0 0 2 1 1 2 2 1 4 3 4 3 2 3 2 1 2 3 3 1 1 1 3 3 4 2 3 2 3 4 2 2 6 5 4 4 5 5 2 2 2 4 2 2\n",
      "C: 3 3 3 3 4 3 2 2 2 3 3 2 5 4 3 2 3 3 2 1 1 3 2 2 1 3 2 2 1 2 1 3 1 2 0 1 2 3 2 5 3 0 1 1 1 5 1 2 1 3 4 2 2 3 3 3 3 4 4 4 3 5 4 2 2 1 1 0 3 1 0 4 2 1 2 3 2 3 3 3 2 3 2 3 1 4 4 3 3 1 4 3 2 4 1 0 5 1 0 2 4 5 1 1 3 1 1 4 1 3 2 2 2 4 3 6 1 4 1 3 1 2 3 1 1 2 4 4 3 2 4 3 3 2 6 4 2 3 2 1 4 2 4 2 2 2 2 0 4 2 1 5 4 0 2 2 0 6 2 3 0 3 2 4 3 1 2 2 3 3 7 4 0 3 1 2 2 2 2 3 2 4 4 1 2 3 1 2 4 1 2 2 3 2 2 3 0 4 2 3 0 0 2 2 3 1 5 4 4 3 3 5 4 3 1 2 3 4 3 3 1 3 3 1 2 4 4 2 2 2 2 3 3 2 1 0 1 3 1 0 3 1 1 2 2 4 3 5 1 3 0 2 0 1 2 2 6 2 1 3 3 5 4 2 5 0 4 2 4 2 2 6 4 2 1 4 3 2 4 5 1 5 3 2 2 6 4 2 2 1 1 2 0 3 1 2 3 3 1 1 3 3 5 5 4 0 1 2 1 2 2 4 3 1 1 3 1 3 3 2 3 2 1 4 2 1 1 3 2 2 5 2 0 1 4 2 0 1 2 3 3 2 3 2 2 1 3 2 2 1 2 2 1 2 3 3 3 2 6 4 2 1 1 3 1 1 2 3 3 1 3 3 7 1 4 2 3 2 1 1 2 5 2 5 4 1 3 1 4 0 3 1 1 4 3 1 3 3 2 4 1 2 0 1 2 2 1 3 1 3 3 1 2 3 0 3 0 2 3 1 1 3 0 1 3 5 0 0 0 4 2 2 3 2 3 1 4 3 1 3 4 0 0 4 4 4 2 2 3 3 6 4 2 3 1 6 0 2 2 4 2 5 2 1 1 2 3 3 2 0 3 2 3 4 2 3 5 1 5 2 4 1 4 4 3 5 3 3 3 1 1 1 1 4 5 3 3 1 1 2 4 3 1 3 2 4 2 2 1 5 2 0 4 1 2 4 4 2 2 2 3 1 2 3 4 1 4 3 1 3 1 4 5 3 3 1 0 2 2 2 5 2 2 2 2 3 4 3 2 1 1 1 1 2 3 4 2 2 2 1 2 3 1 1 3 1 5 1 4 3 2 1 2 4 2 1 3 1 1 2 2 3 4 3 2 2 5 5 4 3 3 0 2 4 3 2 4 1 5 1 2 4 1 2 4 3 1 7 2 2 3 0 2 3 2 1 4 2 0 3 5 2 1 1 2 2 4 3 2 5 4 4 0 4 2 3 4 2 1 3 1 2 2 4 0 2 1 1 4 2 2 5 2 4 4 1 3 3 4 3 4 3 2 1 1 3 3 5 3 4 1 3 3 0 6 2 1 2 5 3 3 1 4 2 1 1 3 2 1 2 5 4 3 0 4 3 2 2 3 2 2 2 3 3 5 1 4 4 1 1 1 0 4 4 5 2 2 4 2 2 2 1 3 3 2 3 2 2 5 1 4 4 4 4 3 4 2 1 2 4 4 3 1 2 3 1 1 3 2 0 2 2 1 3 4 3 4 2 2 2 1 1 2 3 0 1 2 1 2 1 3 0 2 3 3 0 4 5 3 3 1 3 2 3 0 3 1 2 0 1 2 1 6 1 1 4 4 3 0 2 3 2 1 2 4 2 2 1 3 1 1 1 3 2 4 2 1 2 3 4 4 1 1 3 1 6 2 3 2 4 2 2 3 2 3 4 2 2 3 4 3 2 4 3 1 3 2 5 4 2 2 2 2 3 3 4 1 5 3 4 2 1 3 1 1 1 1 0 2 2 4 2 4 4 2 0 2 3 3 3 0 5 2 2 3 0 1 2 4 4 2 5 2 3 2 3 3 3 3 3 3 2 3 2 3 4 2 5 1 3 2 4 4 2 4 3 3 2 2 2 0 1 4 1 2 3 2 2 3 2 2 4 1 4 1 3 2 1 3 5 1 1 3 1 1 1 3 2 2 4 3 4 1 2\n",
      "G: 2 5 4 1 1 0 2 3 4 1 3 2 3 2 2 2 2 2 3 4 3 1 2 3 1 3 3 3 3 2 5 3 4 4 4 3 1 4 2 1 2 2 0 5 3 4 3 3 5 2 3 2 2 4 3 3 5 3 3 1 3 0 2 1 1 3 4 5 3 2 2 1 1 4 1 0 4 2 5 2 2 2 4 3 3 1 2 1 4 4 4 2 1 2 3 2 2 4 4 1 1 2 1 3 2 4 3 1 1 3 5 3 3 1 2 3 0 3 2 1 3 4 0 2 3 3 2 2 1 1 1 4 1 1 0 2 6 2 1 4 1 3 2 3 0 2 3 6 3 3 2 1 1 3 1 3 4 3 2 3 0 1 2 1 1 3 4 1 2 1 0 1 0 2 1 7 4 2 4 1 2 1 2 1 1 5 1 3 3 2 1 2 1 3 1 2 5 2 1 2 1 1 0 5 4 5 2 2 3 2 2 1 0 4 4 2 2 4 3 4 5 0 6 4 1 2 2 4 3 5 3 1 1 1 3 4 3 3 4 4 3 3 2 2 4 1 2 2 5 2 2 3 3 3 1 2 1 4 2 1 0 2 3 4 1 2 1 0 2 1 4 0 1 0 2 1 1 1 1 3 3 1 3 5 2 2 1 4 1 5 1 2 4 4 3 4 2 3 2 4 0 4 2 1 3 5 2 2 1 2 5 3 1 1 1 2 0 1 1 2 3 1 2 2 3 2 4 2 4 3 2 4 1 1 3 4 8 3 2 4 2 2 0 3 3 4 6 3 2 3 0 2 4 3 1 3 3 4 1 2 4 2 2 1 4 2 3 2 2 1 3 3 1 3 4 2 2 6 2 3 2 2 2 2 1 3 3 4 0 4 4 3 3 4 3 2 1 2 4 3 2 5 4 4 3 5 2 1 3 3 2 4 2 3 1 6 2 2 2 3 5 2 5 4 2 3 1 3 2 1 6 2 2 2 2 1 0 2 7 3 2 2 1 0 3 2 2 2 3 3 0 3 2 3 2 1 3 4 1 3 3 2 0 4 2 3 3 3 2 2 1 3 2 3 3 1 1 4 3 3 3 3 2 2 2 2 3 4 0 4 3 4 3 0 2 4 3 2 4 5 1 3 1 3 3 2 4 4 3 0 2 2 3 4 4 1 2 5 2 4 2 3 8 2 2 2 1 3 6 1 2 2 1 0 3 3 4 4 4 2 3 2 3 2 1 2 1 2 4 0 3 5 2 1 2 3 1 3 0 1 3 0 1 3 2 3 1 4 1 4 2 1 4 2 3 3 3 3 2 3 3 2 0 4 6 1 2 1 2 4 5 3 3 1 1 1 1 2 0 3 4 1 3 1 5 3 4 2 4 4 1 1 3 2 1 4 1 2 2 2 4 0 5 4 1 2 1 2 2 1 3 2 3 5 5 1 0 2 3 2 3 2 5 3 5 4 3 3 2 2 1 4 1 1 3 3 2 2 1 1 1 3 2 3 4 1 3 1 2 2 3 2 3 4 1 4 4 5 1 1 2 2 1 1 3 3 3 3 2 1 3 0 2 2 1 1 2 2 2 3 2 4 5 2 2 3 1 2 2 2 3 4 1 2 2 2 2 4 3 2 1 2 3 2 4 3 1 2 1 4 2 2 2 1 2 1 1 5 4 4 1 4 4 3 3 4 5 2 2 7 3 3 3 3 2 2 3 2 3 3 3 1 3 3 2 3 2 1 0 3 1 3 0 4 2 2 0 1 3 1 6 2 2 3 2 1 0 4 3 2 3 0 3 4 5 0 4 3 4 1 2 0 6 4 1 6 3 1 1 2 3 5 3 3 2 3 3 2 1 5 3 3 2 2 2 0 4 3 4 1 2 1 2 4 4 1 2 1 3 1 2 2 3 3 3 2 3 2 1 1 3 1 2 2 2 2 5 2 2 3 2 3 3 2 4 5 3 4 3 4 0 1 4 0 2 4 2 2 2 3 2 3 2 1 4 1 4 5 3 3 1 2 3 5 6 3 3 1 3 4 2 1 2 1 3 1 3 1 3 4 3 3 2 2 2 3 4 4 3 3 1 5 2 3 3 5 3 3 1 1 3 4 2 0 2 3 2 2 2 0 4 4 0 3 4 3 0 3 3 0 2 1 3 3\n",
      "T: 2 0 3 4 3 3 1 1 1 6 3 2 1 2 2 4 1 3 2 5 2 5 3 3 5 3 3 1 4 2 0 2 4 3 2 6 6 1 1 3 1 5 2 2 0 1 2 3 1 3 2 0 2 3 2 1 1 1 3 1 1 3 2 3 4 4 3 1 2 2 5 2 4 5 2 4 2 3 1 3 4 1 1 1 4 2 1 2 1 1 2 4 5 4 2 4 1 3 2 2 0 1 2 4 5 2 4 2 5 4 3 1 2 1 3 1 7 3 1 1 2 1 5 3 4 1 3 1 3 3 4 0 3 3 3 3 1 2 2 1 4 3 3 2 3 3 3 1 1 3 4 4 2 4 0 4 2 1 2 0 5 3 1 2 3 1 3 3 1 3 3 4 8 2 4 0 2 3 2 4 4 3 1 2 3 0 5 2 1 5 3 4 4 3 5 5 2 3 5 3 4 4 6 2 1 3 1 4 2 0 3 4 5 3 1 3 4 2 3 1 2 5 1 3 3 3 3 3 3 1 1 4 3 4 2 4 4 2 3 3 0 4 4 2 2 2 3 1 1 2 2 4 3 2 3 4 0 2 5 3 5 1 1 3 0 4 2 2 3 4 2 1 1 6 3 5 3 4 2 1 1 2 3 2 1 0 3 3 3 4 1 3 4 1 3 1 3 2 4 2 1 1 1 2 1 3 4 1 3 5 2 1 3 4 3 2 4 2 3 3 2 3 3 0 3 4 3 2 2 5 2 3 6 5 2 3 1 2 4 1 2 1 3 2 2 4 0 3 1 4 7 2 2 2 4 1 3 1 1 1 2 6 5 3 2 4 2 1 2 3 2 4 1 2 0 1 3 1 4 4 5 1 4 1 2 2 1 3 6 3 2 4 1 1 2 3 5 2 1 2 3 2 4 4 1 2 5 1 1 3 2 2 3 2 7 0 4 3 3 3 3 1 1 3 4 1 1 5 4 2 2 4 2 6 2 6 2 4 2 1 2 4 2 0 2 1 2 3 2 0 0 2 6 2 5 0 3 2 4 0 1 3 2 3 2 2 2 4 1 5 4 3 2 1 3 1 3 2 0 3 1 2 3 3 3 2 1 2 6 1 2 2 3 2 1 2 2 5 3 2 2 2 3 3 3 3 1 1 1 3 2 6 2 0 1 2 4 2 2 2 1 2 0 3 4 4 3 2 1 3 3 1 0 6 1 4 4 4 2 5 2 4 4 5 2 2 3 3 2 6 4 1 3 4 4 2 1 4 2 2 2 1 5 5 2 3 3 2 2 1 2 6 2 2 3 1 2 1 5 1 2 5 4 2 2 2 2 4 1 3 2 4 2 4 3 3 4 3 3 3 2 4 4 4 1 1 2 1 3 1 5 4 3 1 4 3 1 5 2 4 0 3 3 0 5 1 1 2 3 3 3 3 6 0 2 2 4 4 5 0 2 3 0 2 3 2 4 1 4 4 3 0 2 2 1 2 2 4 1 4 1 2 2 4 2 5 3 2 1 1 2 3 0 3 2 2 2 2 1 2 2 2 0 5 6 0 1 3 4 3 1 2 2 5 2 2 4 4 3 4 5 2 0 1 0 2 3 0 5 4 5 1 3 2 2 2 6 0 2 3 5 4 1 1 1 3 2 5 3 3 0 1 3 2 3 2 3 2 1 1 3 1 3 0 2 3 2 2 0 1 1 3 4 3 2 3 2 2 3 2 3 3 3 2 3 4 5 3 5 2 3 2 6 2 1 3 1 1 3 3 1 2 2 4 3 2 3 0 4 3 3 4 0 3 3 5 1 4 3 3 1 2 2 3 2 2 1 6 3 3 4 3 1 3 4 2 3 3 5 1 1 2 3 2 5 0 4 2 2 2 3 4 3 2 1 1 0 3 2 0 4 4 1 2 3 2 2 2 2 3 5 2 4 4 2 1 1 1 1 2 4 2 2 3 1 1 4 3 2 3 1 3 2 3 1 1 4 4 3 1 3 0 1 4 2 4 2 2 1 2 3 1 4 1 1 1 1 3 1 3 2 5 0 4 3 1 1 3 2 3 5 3 2 5 3 2 1 3 1 2 5 1 2 3 3 1 3 2 3 6 4 1 4 3 3 2 3 5 2 1 3 3 1 1 1 2 2 0 3 4 3 1 4 3\n"
     ]
    }
   ],
   "source": [
    "#Cons(Consensus and Profile)\n",
    "\n",
    "import numpy as np\n",
    "from Bio import SeqIO \n",
    "\n",
    "# Parse the FASTA file\n",
    "with open('rosalind_cons.txt') as input_file:\n",
    "    dna_list = [str(record.seq) for record in SeqIO.parse(input_file, 'fasta')]\n",
    "\n",
    "# Initialize the profile matrix\n",
    "length = len(dna_list[0])\n",
    "profile = np.zeros((4, length), dtype=int)\n",
    "nucleotide_map = {'A': 0, 'C': 1, 'G': 2, 'T': 3}\n",
    "\n",
    "# Fill the profile matrix\n",
    "for dna in dna_list:\n",
    "    for index, nucleotide in enumerate(dna):\n",
    "        profile[nucleotide_map[nucleotide]][index] += 1\n",
    "\n",
    "# Determine the consensus string\n",
    "consensus = ''\n",
    "nucleotide_order = ['A', 'C', 'G', 'T']\n",
    "for i in range(length):\n",
    "    most_common = np.argmax(profile[:, i])\n",
    "    consensus += nucleotide_order[most_common]\n",
    "\n",
    "output = [consensus]\n",
    "for i, nucleotide in enumerate(nucleotide_order):\n",
    "    output.append(f\"{nucleotide}: {' '.join(map(str, profile[i]))}\")\n",
    "\n",
    "print('\\n'.join(output)) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 4\n",
      "20 4\n",
      "23 4\n",
      "30 4\n",
      "57 4\n",
      "89 4\n",
      "95 4\n",
      "98 4\n",
      "101 4\n",
      "142 4\n",
      "149 4\n",
      "160 4\n",
      "170 4\n",
      "176 4\n",
      "182 4\n",
      "183 4\n",
      "196 4\n",
      "239 4\n",
      "269 4\n",
      "272 4\n",
      "274 4\n",
      "276 4\n",
      "295 4\n",
      "304 4\n",
      "306 4\n",
      "311 4\n",
      "328 4\n",
      "332 4\n",
      "340 4\n",
      "343 4\n",
      "359 4\n",
      "368 4\n",
      "380 4\n",
      "409 4\n",
      "460 4\n",
      "464 4\n",
      "488 4\n",
      "494 4\n",
      "532 4\n",
      "575 4\n",
      "603 4\n",
      "607 4\n",
      "619 4\n",
      "623 4\n",
      "626 4\n",
      "666 4\n",
      "680 4\n",
      "691 4\n",
      "702 4\n",
      "709 4\n",
      "720 4\n",
      "726 4\n",
      "728 4\n",
      "743 4\n",
      "750 4\n",
      "756 4\n",
      "768 4\n",
      "788 4\n",
      "789 4\n",
      "793 4\n",
      "802 4\n",
      "813 4\n",
      "817 4\n",
      "19 6\n",
      "22 6\n",
      "94 6\n",
      "169 6\n",
      "175 6\n",
      "238 6\n",
      "273 6\n",
      "327 6\n",
      "342 6\n",
      "367 6\n",
      "408 6\n",
      "463 6\n",
      "493 6\n",
      "602 6\n",
      "606 6\n",
      "679 6\n",
      "690 6\n",
      "725 6\n",
      "727 6\n",
      "812 6\n",
      "272 8\n",
      "407 8\n",
      "462 8\n",
      "811 8\n",
      "406 10\n",
      "405 12\n"
     ]
    }
   ],
   "source": [
    "#Revp (Locating Restriction Sites)\n",
    "# Function to find reverse palindromes in the DNA sequence\n",
    "# Palindromes of length between 4 and 12\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "with open('rosalind_revp.txt') as input_file:\n",
    "    dna = str(next(SeqIO.parse(input_file, 'fasta')).seq)\n",
    "\n",
    "locations = []\n",
    "\n",
    "\n",
    "for length in range(4, 13):  \n",
    "    for index in range(len(dna) - length + 1):\n",
    "        segment = dna[index:index + length]\n",
    "        if segment == str(Seq(segment).reverse_complement()):\n",
    "            locations.append(f\"{index + 1} {length}\")\n",
    "\n",
    "print(\"\\n\".join(locations))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MDLLANQPCGTVGGAGPPGLDCQSYERDRVVTRAIRIPSSGRFDRYNGIKHGWRSLLGLLYLLRHGRPAWRHESSSFLDASTLIIPYIFRTYKKYNKAEAPTTNKLSPVNASRFDGCMALTLRTFSPAASSTSLRGAGHLILYRYITCQSKFILRSYRRSTARFLPTQLWRTRLLPETSRVGQET\n"
     ]
    }
   ],
   "source": [
    "#SPLC (RNA Splicing)\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "with open('rosalind_splc.txt') as input_file:\n",
    "    sequences = [str(record.seq) for record in SeqIO.parse(input_file, \"fasta\")]\n",
    "#initialiazation :\n",
    "main_dna = sequences[0] #pre-mRNA is the first \n",
    "introns = sequences[1:] \n",
    "#get the introns out off the sequence\n",
    "for intron in introns:\n",
    "    main_dna = main_dna.replace(intron, '')\n",
    "\n",
    "# Translate the mature DNA sequence to a protein sequence (via Translate function)\n",
    "protein = Seq(main_dna).translate(to_stop=True)\n",
    "print(protein)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.103448275862069\n"
     ]
    }
   ],
   "source": [
    "#TRAN (Transitions and Transversions)\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "with open('rosalind_tran.txt') as input_file:\n",
    "    sequences = [str(record.seq) for record in SeqIO.parse(input_file, \"fasta\")]\n",
    "\n",
    "#Initialization :\n",
    "dna1, dna2 = sequences[0], sequences[1]\n",
    "purines = {'A', 'G'}\n",
    "pyrimidines = {'C', 'T'}\n",
    "transitions = 0\n",
    "transversions = 0\n",
    "\n",
    "# Comparing the two DNA sequences \n",
    "for base1, base2 in zip(dna1, dna2):\n",
    "    if base1 != base2:\n",
    "        if (base1 in purines and base2 in purines) or (base1 in pyrimidines and base2 in pyrimidines):\n",
    "            transitions += 1\n",
    "        else:\n",
    "            transversions += 1\n",
    "\n",
    "print(transitions / transversions)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
